Add CustomResourceDefinition for Instance Manager

ttpcodes opened the pull request:
This PR creates a new `Custom Resource Definition` called the `Instance Manager`, intended to be an abstraction of the `Engine Process Pods` and `Replica Process Pods` that will be created on each `Node` as part of the new `engine launcher` refactor. 

The `longhorn/v1alpha1` API has been updated accordingly to generate the new definitions, and the `DataStore` has been updated accordingly to handle `Instance Managers` and have methods to interact with them.

shuo-wu commented on types/resource.go:
```diff
@@ -282,3 +282,51 @@ type BackupStatus struct {
+)
+
+type InstanceManagerSpec struct {
+	BootNodeID  string `json:"bootNodeID"`
```
Rename it to `NodeBootID`. And should be it in here rather than `InstanceManagerStatus`?

ttpcodes replied:
Will fix.

ttpcodes commented:
- Added a generic listing function for all `Instance Managers`.
- Added a listing function to get all the `Instance Managers` that exist on a particular `Node`.
- Renamed `BootNodeID` to `NodeBootID` and moved it to `Instance Manager Status`.
- Added missing JSON fields for `Instance Process Status`.

yasker commented on types/resource.go:
```diff
@@ -282,3 +282,51 @@ type BackupStatus struct {
+	EngineImage string `json:"engineImage"`
+	NodeID      string `json:"nodeID"`
+	OwnerID     string `json:"ownerID"`
+	Type        InstanceManagerType
```
Missing json field.

ttpcodes replied:
Adding JSON field.

yasker commented on types/resource.go:
```diff
@@ -282,3 +282,51 @@ type BackupStatus struct {
+
+type InstanceManagerStatus struct {
+	CurrentState InstanceManagerState `json:"currentState"`
+	Instances    map[string]InstanceProcessStatus
```
Missing json field.

ttpcodes replied:
Adding JSON field.

yasker commented on k8s/pkg/apis/longhorn/v1alpha1/zz_generated.deepcopy.go:
```diff
@@ -146,6 +146,67 @@ func (in *EngineList) DeepCopyObject() runtime.Object {
 	return nil
 }
 
+// DeepCopyInto is an autogenerated deepcopy function, copying the receiver, writing into out. in must be non-nil.
```
Separates added code and generated code.

ttpcodes replied:
Separating manually added code versus generated code.

shuo-wu commented on types/resource.go:
```diff
@@ -282,3 +282,51 @@ type BackupStatus struct {
+	Type        InstanceManagerType
+}
+
+type InstanceManagerStatus struct {
```
Missing field `IP`

ttpcodes replied:
Adding.

ttpcodes commented:
- Fixed a few remaining missing `JSON` fields.
- Split the initial `Custom Resource Definition` commit into user-defined code and `client-gen` outputted code.
- Added missing `IP` field.
